Sources/PackageCollections/JSONModel/JSONCollection+v1.swift

```diff
@@ -252,115 +249,67 @@ extension JSONPackageCollectionModel.V1 {
-                    break
-                }
-
-                minorCount += 1
-            }
+extension JSONPackageCollectionModel.V1.ProductType: Codable {
+    private enum CodingKeys: String, CodingKey {
+        case library, executable, test
+    }
 
-            package.versions.forEach { version in
-                if version.products.isEmpty {
-                    messages.append(.error("Package \(packageID) version \(version.version) does not contain any products.", property: "version.products"))
-                }
-                version.products.forEach { product in
-                    if product.targets.isEmpty {
-                        messages.append(.error("Product \(product.name) of package \(packageID) version \(version.version) does not contain any targets.", property: "product.targets"))
-                    }
-                }
-
-                if version.targets.isEmpty {
-                    messages.append(.error("Package \(packageID) version \(version.version) does not contain any targets.", property: "version.targets"))
-                }
-            }
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        switch self {
+        case let .library(a1):
+            var unkeyedContainer = container.nestedUnkeyedContainer(forKey: .library)
+            try unkeyedContainer.encode(a1)
+        case .executable:
+            try container.encodeNil(forKey: .executable)
+        case .test:
+            try container.encodeNil(forKey: .test)
         }
+    }
 
-        public struct Configuration {
-            public var maximumPackageCount: Int
-            public var maximumMajorVersionCount: Int
-            public var maximumMinorVersionCount: Int
-
-            public init(maximumPackageCount: Int? = nil,
-                        maximumMajorVersionCount: Int? = nil,
-                        maximumMinorVersionCount: Int? = nil) {
-                // TODO: where should we read defaults from?
-                self.maximumPackageCount = maximumPackageCount ?? 50
-                self.maximumMajorVersionCount = maximumMajorVersionCount ?? 2
-                self.maximumMinorVersionCount = maximumMinorVersionCount ?? 3
-            }
+    public init(from decoder: Decoder) throws {
+        let values = try decoder.container(keyedBy: CodingKeys.self)
+        guard let key = values.allKeys.first(where: values.contains) else {
+            throw DecodingError.dataCorrupted(.init(codingPath: decoder.codingPath, debugDescription: "Did not find a matching key"))
+        }
+        switch key {
+        case .library:
+            var unkeyedValues = try values.nestedUnkeyedContainer(forKey: key)
+            let a1 = try unkeyedValues.decode(Self.LibraryType.self)
+            self = .library(a1)
+        case .test:
+            self = .test
+        case .executable:
+            self = .executable
```

Comment:
The diff obscures this, but this copies `ProductType` (and its `Codable` conformance) to a new type `JSONPackageCollectionModel.V1.ProductType`.